extended/src/main/java/io/kubernetes/client/extended/controller/DefaultController.java

```diff
@@ -0,0 +1,235 @@
+    this.workQueue = workQueue;
+    this.readyFuncs = readyFuncs;
+    this.readyTimeout = Duration.ofSeconds(30);
+    this.readyCheckInternal = Duration.ofSeconds(1);
+  }
+
+  // preFlightCheck checks if the controller is ready for working.
+  private boolean preFlightCheck() {
+    if (workerCount <= 0) {
+      log.error("Fail to start controller {}: worker count must be positive.");
+      return false;
+    }
+    if (workerThreadPool == null) {
+      log.error("Fail to start controller {}: missing worker thread-pool.");
+      return false;
+    }
+    if (!isReady()) {
+      log.error("Timed out waiting for cache to be synced.");
+      return false;
+    }
+    return true;
+  }
+
+  // isReady checks if any ready functions complaining before working.
+  private boolean isReady() {
+    boolean synced = true;
+    if (this.readyFuncs.length > 0) {
+      synced =
+          Wait.poll(
+              readyCheckInternal,
+              readyTimeout,
+              () -> {
+                boolean ready = true;
+                for (Supplier<Boolean> cacheHasSyncedFunc : this.readyFuncs) {
+                  ready = ready && cacheHasSyncedFunc.get();
+                }
+                return ready;
+              });
+    }
+    return synced;
+  }
+
+  @Override
+  public void run() {
+    if (!preFlightCheck()) {
+      log.error("Controller {} failed pre-run check, exiting..", this.name);
+      return;
+    }
+
+    // spawns worker threads for the controller.
+    CountDownLatch latch = new CountDownLatch(workerCount);
+    for (int i = 0; i < this.workerCount; i++) {
+      final int workerIndex = i;
+      workerThreadPool.scheduleWithFixedDelay(
+          () -> {
+            if (log.isDebugEnabled()) {
+              log.debug("Starting controller {} worker {}..", this.name, workerIndex);
+            }
+            this.worker();
+            latch.countDown();
+            if (log.isDebugEnabled()) {
+              log.debug("Exiting controller {} worker {}..", this.name, workerIndex);
+            }
+          },
+          0,
+          1,
+          TimeUnit.SECONDS);
+    }
+    try {
+      if (log.isDebugEnabled()) {
+        log.debug("Controller {} bootstrapping..", this.name);
+      }
+      latch.await();
+    } catch (Throwable t) {
+      log.error("Controller {} hitting exception.", this.name);
+      log.error("Aborting controller.", t);
+    } finally {
+      log.info("Controller {} exited", this.name);
+    }
+  }
+
+  @Override
+  public void shutdown() {
+    // shutdown work-queue before the thread-pool
+    workQueue.shutDown();
+    workerThreadPool.shutdown();
+  }
+
+  private void worker() {
+    // taking tasks from work-queue in a loop
+    while (!workQueue.isShuttingDown()) {
+      Reconciler.Request request = null;
+      try {
+        request = workQueue.get();
+      } catch (InterruptedException e) {
+        // we're reaching here mostly because of forcibly shutting down the controller.
```

Comment:
I think alignment is messed up?

Reply:
no? i think? the formatter requests it..